transports/onion/src/address.rs

```diff
@@ -0,0 +1,107 @@
+use std::net::{IpAddr, SocketAddr};
+
+use arti_client::{IntoTorAddr, TorAddr, TorAddrError};
+use libp2p_core::{multiaddr::Protocol, Multiaddr};
+
+pub fn dangerous_extract_tor_address(multiaddr: &Multiaddr) -> Result<TorAddr, TorAddrError> {
+    let socket_addr = try_extract_socket_addr(&mut multiaddr.clone())?;
+    TorAddr::dangerously_from(socket_addr)
+}
+
+pub fn safe_extract_tor_address(multiaddr: &mut Multiaddr) -> Result<TorAddr, TorAddrError> {
+    let tcp_variant = multiaddr.pop().ok_or(TorAddrError::NoPort)?;
+    let tcp_port = match tcp_variant {
+        Protocol::Tcp(p) => p,
+        _ => return Err(TorAddrError::NoPort),
+    };
+    let dns_variant = multiaddr.pop().ok_or(TorAddrError::InvalidHostname)?;
+    let dns_addr = match dns_variant {
+        Protocol::Dns(dns) => dns,
+        Protocol::Dns4(dns) => dns,
+        Protocol::Dns6(dns) => dns,
+        _ => return Err(TorAddrError::InvalidHostname),
+    };
+    let address_tuple = (dns_addr.as_ref(), tcp_port);
+    address_tuple.into_tor_addr()
+}
+
+fn try_extract_socket_addr(multiaddr: &mut Multiaddr) -> Result<SocketAddr, TorAddrError> {
+    let tcp_variant = multiaddr.pop().ok_or(TorAddrError::NoPort)?;
+    let port = match tcp_variant {
+        Protocol::Tcp(p) => p,
+        _ => return Err(TorAddrError::NoPort),
+    };
+    let ip_variant = multiaddr.pop().ok_or(TorAddrError::InvalidHostname)?;
+    let ip = match ip_variant {
+        Protocol::Ip4(ip4) => IpAddr::V4(ip4),
+        Protocol::Ip6(ip6) => IpAddr::V6(ip6),
+        _ => return Err(TorAddrError::InvalidHostname),
+    };
+    Ok(SocketAddr::new(ip, port))
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{borrow::Cow, net::Ipv4Addr, ops::Not};
+
+    use libp2p_core::{multiaddr::multiaddr, Multiaddr};
+
+    use super::safe_extract_tor_address;
+
+    #[test]
+    fn extract_correct_address() {
+        let dns_test_addr = multiaddr!(Dns(Cow::Borrowed("ip.tld")), Tcp(10u16));
```

Comment:
You may find that constructing the mutliaddress from a string and parsing it is easier to read, i.e. this line could be written as:

```suggestion
        let dns_test_addr = "/dns/ip.tld/tcp/10".parse::<Multiaddr>().unwrap();
```

Reply:
I agree, I just always wanted to try that macro 🙃